File: model_1/CS3233_2023_Midterm/indistinguishable_sequences_1_solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

const long long MOD = 1000003233;
const int MAXN = 100100;

class FenwMax {
public:
    vector<int> fenw;
    int n;
    FenwMax(int size) {
        n = size;
        fenw.assign(n+1, 0);
    }

    void update(int idx, int val) {
        while (idx <= n) {
            if (val > fenw[idx]) {
                fenw[idx] = val;
            }
            idx += (idx & -idx);
        }
    }

    int query(int idx) {
        if (idx <= 0) return 0;
        int res = 0;
        while (idx > 0) {
            if (fenw[idx] > res) 
                res = fenw[idx];
            idx -= (idx & -idx);
        }
        return res;
    }
};

class FenwSum {
public:
    vector<long long> fenw;
    int n;
    FenwSum(int size) {
        n = size;
        fenw.assign(n+1, 0);
    }

    void update(int idx, long long delta) {
        while (idx <= n) {
            fenw[idx] = (fenw[idx] + delta) % MOD;
            idx += (idx & -idx);
        }
    }

    long long query(int idx) {
        long long res = 0;
        while (idx > 0) {
            res = (res + fenw[idx]) % MOD;
            idx -= (idx & -idx);
        }
        return res;
    }

    long long range_query(int l, int r) {
        if (l > r) return 0;
        long long left_part = (l-1>0) ? query(l-1) : 0;
        long long right_part = query(r);
        return (right_part - left_part + MOD) % MOD;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    int n;
    cin >> n;
    vector<int> a(n);
    for (int i=0; i<n; i++) {
        cin >> a[i];
    }

    vector<long long> pow2(MAXN, 1);
    for (int i=1; i<MAXN; i++) {
        pow2[i] = (pow2[i-1] * 2) % MOD;
    }

    long long inv2 = (MOD+1)/2;
    vector<long long> inv_pow2(MAXN, 1);
    for (int i=1; i<MAXN; i++) {
        inv_pow2[i] = (inv_pow2[i-1] * inv2) % MOD;
    }

    vector<int> f(n, 1);
    FenwMax fenw_f(n);
    for (int i=0; i<n; i++) {
        if (a[i] > 1) {
            int q = fenw_f.query(a[i]-1);
            f[i] = q + 1;
        } else {
            f[i] = 1;
        }
        fenw_f.update(a[i], f[i]);
    }

    vector<int> g(n, 1);
    FenwMax fenw_g(n);
    for (int i=n-1; i>=0; i--) {
        int b_val = n+1 - a[i];
        if (b_val > 1) {
            int q = fenw_g.query(b_val-1);
            g[i] = q + 1;
        } else {
            g[i] = 1;
        }
        fenw_g.update(b_val, g[i]);
    }

    int L = 0;
    for (int i=0; i<n; i++) {
        if (f[i] > L) L = f[i];
    }

    vector<int> F;
    for (int i=0; i<n; i++) {
        if (f[i] + g[i] - 1 == L) {
            F.push_back(i);
        }
    }

    vector< vector<int> > levels(L+1);
    for (int i : F) {
        int k = f[i];
        if (k <= L) {
            levels[k].push_back(a[i]);
        }
    }

    long long dp0 = 1;
    vector<long long> dp(n+1, 0);

    for (int k=1; k<=L; k++) {
        vector<int> A = levels[k];
        if (A.empty()) {
            continue;
        }
        int total = A.size();
        vector<int> prefix_count_arr(n+1, 0);
        for (int v : A) {
            if (v >= 1 && v <= n) {
                prefix_count_arr[v] = 1;
            }
        }
        for (int i=1; i<=n; i++) {
            prefix_count_arr[i] += prefix_count_arr[i-1];
        }

        vector<long long> F_arr(n+1, 0);
        for (int s=1; s<=n; s++) {
            F_arr[s] = dp[s] * pow2[prefix_count_arr[s]] % MOD;
        }

        FenwSum fenw_tree(n);
        for (int s=1; s<=n; s++) {
            fenw_tree.update(s, F_arr[s]);
        }

        vector<long long> next_dp_nonempty(n+1, 0);
        for (int v : A) {
            long long term0 = dp0;
            term0 = term0 * pow2[total - prefix_count_arr[v]] % MOD;

            long long sum_range = fenw_tree.range_query(v, n);
            long long term_positive = sum_range * inv_pow2[prefix_count_arr[v]] % MOD;

            next_dp_nonempty[v] = (term0 + term_positive) % MOD;
        }

        long long new_dp0 = dp0;
        vector<long long> new_dp = dp;
        for (int v : A) {
            new_dp[v] = (new_dp[v] + next_dp_nonempty[v]) % MOD;
        }

        dp0 = new_dp0;
        dp = new_dp;
    }

    long long valid_subsets = dp0;
    for (int i=1; i<=n; i++) {
        valid_subsets = (valid_subsets + dp[i]) % MOD;
    }

    long long non_critical_count = n - F.size();
    long long total_subsets = pow2[n];
    long long ans = (total_subsets - pow2[non_critical_count] * valid_subsets % MOD + MOD) % MOD;
    cout << ans << endl;

    return 0;
}

File: model_1/CS3233_2023_Midterm/indistinguishable_sequences_1_solution.py
MOD = 1000003233

def main():
    import sys
    data = sys.stdin.read().split()
    if not data:
        return
    n = int(data[0])
    a = list(map(int, data[1:1+n]))
    
    if n == 0:
        print(0)
        return

    size = n
    fenw_f = [0] * (size + 1)
    
    def fenw_update_max(fenw, size, i, val):
        while i <= size:
            if val > fenw[i]:
                fenw[i] = val
            i += i & -i

    def fenw_query_max(fenw, i):
        res = 0
        while i:
            if fenw[i] > res:
                res = fenw[i]
            i -= i & -i
        return res

    f_arr = [1] * n
    for i in range(n):
        val = a[i]
        if val > 1:
            q = fenw_query_max(fenw_f, val - 1)
        else:
            q = 0
        f_arr[i] = q + 1
        fenw_update_max(fenw_f, size, val, f_arr[i])
    
    c = [size + 1 - x for x in a]
    fenw_g = [0] * (size + 1)
    g_arr = [1] * n
    for i in range(n - 1, -1, -1):
        val = c[i]
        if val > 1:
            q = fenw_query_max(fenw_g, val - 1)
        else:
            q = 0
        g_arr[i] = q + 1
        fenw_update_max(fenw_g, size, val, g_arr[i])
    
    L_val = max(f_arr) if n > 0 else 0
    
    F_indices = []
    for i in range(n):
        if f_arr[i] + g_arr[i] - 1 == L_val:
            F_indices.append(i)
    total_critical = len(F_indices)
    non_critical = n - total_critical

    levels = [[] for _ in range(L_val + 1)]
    for i in F_indices:
        lev = f_arr[i]
        if lev <= L_val:
            levels[lev].append(a[i])
    
    for k in range(1, L_val + 1):
        levels[k].sort(reverse=True)
    
    pow2 = [1] * (n + 1)
    for i in range(1, n + 1):
        pow2[i] = (pow2[i - 1] * 2) % MOD
        
    inv2 = pow(2, MOD - 2, MOD)
    inv_pow2 = [1] * (n + 1)
    for i in range(1, n + 1):
        inv_pow2[i] = (inv_pow2[i - 1] * inv2) % MOD

    dp0 = 1
    dp = [0] * (n + 1)
    
    for k in range(1, L_val + 1):
        level_vals = levels[k]
        total_in_level = len(level_vals)
        if total_in_level == 0:
            continue
            
        cnt_prefix_arr = [0] * (n + 1)
        for val in level_vals:
            if 1 <= val <= n:
                cnt_prefix_arr[val] += 1
        for i in range(1, n + 1):
            cnt_prefix_arr[i] += cnt_prefix_arr[i - 1]
        
        F_positive = [0] * (n + 1)
        for s in range(1, n + 1):
            F_positive[s] = dp[s] * pow2[cnt_prefix_arr[s]] % MOD
        
        suffix_sum = [0] * (n + 2)
        for s in range(n, 0, -1):
            suffix_sum[s] = (suffix_sum[s + 1] + F_positive[s]) % MOD

        next_dp_level = [0] * (n + 1)
        for v in level_vals:
            if v < 1 or v > n:
                continue
            term0 = dp0 * pow2[total_in_level - cnt_prefix_arr[v]] % MOD
            term_positive = suffix_sum[v] * inv_pow2[cnt_prefix_arr[v]] % MOD
            next_dp_level[v] = (term0 + term_positive) % MOD
        
        new_dp0 = dp0
        new_dp = [0] * (n + 1)
        for s in range(1, n + 1):
            new_dp[s] = dp[s]
        for v in level_vals:
            if 1 <= v <= n:
                new_dp[v] = (new_dp[v] + next_dp_level[v]) % MOD
        
        dp0 = new_dp0
        dp = new_dp

    total_valid = dp0
    for i in range(1, n + 1):
        total_valid = (total_valid + dp[i]) % MOD

    total_subsets = pow2[n]
    non_critical_power = pow2[non_critical]
    result = (total_subsets - non_critical_power * total_valid) % MOD
    if result < 0:
        result += MOD
    print(result)

if __name__ == '__main__':
    main()